### ISSUE TITLE
Finalizer of InputDevice can throw an exception

### ISSUE THREAD

**Issue by MuElhusseiny** [melanchall/drywetmidi#91]
works fine but when my midi keyboard disconnected for any reason and connect it again when i make connect program crashed 
![Capture](https://user-images.githubusercontent.com/69313404/90794182-c3a98180-e30c-11ea-9e79-58c3931fb25f.PNG)

**Comment by melanchall** [comment]
Hi again,

Please try to wrap `Dispose` to `try`-`catch` block:

```csharp
try
{
    if (_inputDevice == null)
        _inputDevice.Dispose();
}
catch (MidiDeviceException) { }
```

**Comment by MuElhusseiny** [comment]
![Capture](https://user-images.githubusercontent.com/69313404/90796191-47646d80-e30f-11ea-994c-ab5865520c28.PNG)

**Comment by melanchall** [comment]
Of course I meant `!= null` in `if`, sorry.

**Comment by MuElhusseiny** [comment]
same problem

**Comment by MuElhusseiny** [comment]
but after some times

**Comment by melanchall** [comment]
What problem? `MidiDeviceException` or `NullReferenceException`? Please show the code and where an exception occured.

**Comment by MuElhusseiny** [comment]
![Capture](https://user-images.githubusercontent.com/69313404/90799403-7977ce80-e313-11ea-8b82-7a3ada7409bd.PNG)

**Comment by melanchall** [comment]
That's OK. You see an exception in debug, but in Release you won't. Just press Continue button in the dialog. Exception will be swallowed by the `catch` block.

**Comment by MuElhusseiny** [comment]
ok im very thanks i will try

**Comment by MuElhusseiny** [comment]
when i select release 

using Melanchall.DryWetMidi.Devices;

could not be found

**Comment by melanchall** [comment]
Please show the error.

**Comment by MuElhusseiny** [comment]
![Capture](https://user-images.githubusercontent.com/69313404/90801060-b93fb580-e315-11ea-9030-ebdbea3e9d05.PNG)

**Comment by melanchall** [comment]
On the screenshot the library is with warning sign at References section at right. How did you put a reference to the library there? Via NuGet?

**Comment by MuElhusseiny** [comment]
no i downloaded dll file and import it to references

**Comment by MuElhusseiny** [comment]
fixed by when clicked on refresh button to update list box same times got same error 

        private void keyref_Click(object sender, RoutedEventArgs e)
        {
                   
            Keyboard.Items.Clear();
            for (int device = 0; device < MidiIn.NumberOfDevices; device++)
            {

                Keyboard.Items.Add(MidiIn.DeviceInfo(device).ProductName);
                Keyboard.Items.Remove("Keyboard");
                Keyboard.Items.Remove("rhythms");
                Keyboard.Items.Remove("WavesMusic1");
                Keyboard.Items.Remove("WavesMusic2");
                Keyboard.Items.Remove("WavesMusic3");
                Keyboard.Items.Remove("Microsoft GS Wavetable Synth");
            }

   
 
        }

**Comment by melanchall** [comment]
What error do you have?

**Comment by MuElhusseiny** [comment]
![Capture](https://user-images.githubusercontent.com/69313404/90813869-7cc98500-e328-11ea-98e6-78dd28a6db66.PNG)

when refresh button clicked 

    private void keyref_Click(object sender, RoutedEventArgs e)
    {
               
        Keyboard.Items.Clear();
        for (int device = 0; device < MidiIn.NumberOfDevices; device++)
        {

            Keyboard.Items.Add(MidiIn.DeviceInfo(device).ProductName);
            Keyboard.Items.Remove("Keyboard");
            Keyboard.Items.Remove("rhythms");
            Keyboard.Items.Remove("WavesMusic1");
            Keyboard.Items.Remove("WavesMusic2");
            Keyboard.Items.Remove("WavesMusic3");
            Keyboard.Items.Remove("Microsoft GS Wavetable Synth");
        }



    }

**Comment by melanchall** [comment]
I see that you're again in debugger. If you run the program without debugger (juzt double ckick exe file of your program), you won't get the exception. When you're in debugger, just click Continue button.

**Comment by MuElhusseiny** [comment]
when i run exe program gor crashed and closed

**Comment by melanchall** [comment]
Hmm, pretty strange. Can you send me entire solution?

**Comment by MuElhusseiny** [comment]
[code.txt](https://github.com/melanchall/drywetmidi/files/5105203/code.txt)

full xaml.cs

**Comment by melanchall** [comment]
I can't run this code without XAML and other files. Can you please create zip archive of entire folder where sln file placed? So I can open solution in VS and run.

**Comment by MuElhusseiny** [comment]
in ur mail

**Comment by melanchall** [comment]
There is no file on my mail.

**Comment by MuElhusseiny** [comment]
107MB My internet to slow 

can you sent to me a catch if no connected device do noting i think error here

**Comment by MuElhusseiny** [comment]
or auto add items to list box if device connected and remove items if device removed..

**Comment by melanchall** [comment]
You can always get all connected devices via `InputDevice.GetAll()` and `OutputDevice.GetAll()`. If there are no devices, do nothing,

**Comment by melanchall** [comment]
During conversation by email we identified that finalizer of `InputDevice` can throw an exception which crashes entire app. It's a bug of DryWetMIDI and it will be fixed.

**Comment by melanchall** [comment]
Fixed by 098d100db7334469fdffa2c077ce1de6ccf860e9.
